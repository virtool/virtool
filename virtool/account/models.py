from datetime import datetime

from email_validator import EmailSyntaxError, validate_email
from pydantic import ConstrainedStr, validator

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.enums import AnalysisWorkflow
from virtool_core.models.group import GroupMinimal, Permissions
from virtool_core.models.user import User


def check_email(email: str | None) -> str | None:
    """Checks if the given email is valid."""
    if email is None:
        return None

    try:
        validate_email(email)
    except EmailSyntaxError:
        raise ValueError("The format of the email is invalid")

    return email


class AccountSettings(BaseModel):
    quick_analyze_workflow: AnalysisWorkflow
    show_ids: bool
    show_versions: bool
    skip_quick_analyze_dialog: bool


class ConstrainedEmail(ConstrainedStr):
    strip_whitespace = True


class Account(User):
    settings: AccountSettings
    email: ConstrainedEmail | None

    _email_validation = validator("email", allow_reuse=True)(check_email)


class APIKey(BaseModel):
    id: str
    created_at: datetime
    groups: list[GroupMinimal]
    name: str
    permissions: Permissions
