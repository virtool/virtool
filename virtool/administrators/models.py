from typing import List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import UserNested


class AdministratorMinimal(UserNested):
    role: AdministratorRole


class Administrator(AdministratorMinimal):
    available_roles: List[dict]


class AdministratorSearch(BaseModel):
    items: List[AdministratorMinimal]
    available_roles: List[dict]
