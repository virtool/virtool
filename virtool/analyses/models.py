import enum
from sqlalchemy import Column, Integer, String, DateTime, Enum

from virtool.postgres import Base


class AnalysisFormat(str, enum.Enum):
    """
    Enumerated type for analysis file formats

    """

    sam = "sam"
    bam = "bam"
    fasta = "fasta"
    fastq = "fastq"
    csv = "csv"
    tsv = "tsv"
    json = "json"


class AnalysisFile(Base):
    """
    SQL model to store new analysis files

    """
    __tablename__ = "analysis_files"

    id = Column(Integer, primary_key=True)
    analysis = Column(String)
    description = Column(String)
    format = Column(Enum(AnalysisFormat))
    name = Column(String)
    name_on_disk = Column(String, unique=True)
    size = Column(Integer)
    uploaded_at = Column(DateTime)

    def __repr__(self):
        return f"<AnalysisFile(id={self.id}, analysis={self.analysis}, description={self.description}, " \
               f"format={self.format}, name={self.name}, name_on_disk={self.name_on_disk}, size={self.size}, " \
               f"uploaded_at={self.uploaded_at})>"
