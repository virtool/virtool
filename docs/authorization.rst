Authorization
=============

.. automodule:: virtool.authorization.__init__
    :members:

Client
------

.. automodule:: virtool.authorization.client
    :members:

