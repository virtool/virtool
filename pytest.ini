[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
