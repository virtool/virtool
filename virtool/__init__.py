"""Application module.

TODO: Remove import when spaces are completely removed.
"""

import virtool.spaces.sql  # noqa: F401
