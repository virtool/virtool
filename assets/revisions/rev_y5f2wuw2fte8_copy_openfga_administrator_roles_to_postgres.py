"""copy openfga administrator roles to postgres

Revision ID: y5f2wuw2fte8
Date: 2025-09-15 21:35:21.415101

"""

import arrow
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from virtool.migration import MigrationContext

# Revision identifiers.
name = "copy openfga administrator roles to postgres"
created_at = arrow.get("2025-09-15 21:35:21.415101")
revision_id = "y5f2wuw2fte8"

alembic_down_revision = "c4d8f879657f"
virtool_down_revision = None

# Change this if an Alembic revision is required to run this migration.
required_alembic_revision = None


async def upgrade(ctx: MigrationContext) -> None:
    """Copy administrator roles from OpenFGA to Postgres."""
    # Get all users from Postgres
    async with AsyncSession(ctx.pg) as session:
        result = await session.execute(
            text("SELECT id, legacy_id FROM users WHERE legacy_id IS NOT NULL")
        )
        users = result.fetchall()

    # For each user, get their administrator role from OpenFGA and update Postgres
    for user in users:
        user_id, role = await ctx.authorization.get_administrator(user.legacy_id)

        if role is not None:
            async with AsyncSession(ctx.pg) as session:
                await session.execute(
                    text("UPDATE users SET administrator_role = :role WHERE id = :id"),
                    {"role": role.value, "id": user.id},
                )
                await session.commit()
