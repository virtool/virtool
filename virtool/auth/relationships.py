from typing import List

from virtool_core.models.enums import Permission

from virtool.data.errors import ResourceNotFoundError
from virtool.errors import DatabaseError
from virtool.groups.db import update_member_users, get_merged_permissions
from virtool.mongo.core import DB
from virtool.mongo.utils import get_one_field, id_exists
from virtool.users.db import compose_groups_update


class BaseRelationship:
    def __init__(self, relations):
        self.object_name = "virtool"
        self.object_type = "app"
        self.user_type = "user"
        self.relations = relations

    async def add(self, mongo: DB):
        ...

    async def remove(self, mongo: DB):
        ...


class GroupMemberships(BaseRelationship):
    def __init__(self, user_id, group_id, relations: List):
        super().__init__(relations)
        self.user_id = user_id
        self.object_name = group_id
        self.object_type = "group"

    async def add(self, mongo: DB):
        groups = await get_one_field(mongo.users, "groups", {"_id": self.user_id}) or []

        try:
            update = await compose_groups_update(mongo, [self.object_name, *groups])
        except DatabaseError as err:
            raise ResourceNotFoundError(str(err))

        await mongo.users.update_one(
            {"_id": self.user_id}, {"$set": update}
        )

    async def remove(self, mongo: DB):
        groups = await get_one_field(mongo.users, "groups", {"_id": self.user_id}) or []

        if self.object_name not in groups:
            raise ResourceNotFoundError

        groups.remove(self.object_name)

        update = {"groups": groups, "permissions": await get_merged_permissions(mongo, groups)}

        await mongo.users.update_one(
            {"_id": self.user_id}, {"$set": update}
        )


class GroupPermissions(BaseRelationship):
    def __init__(self, group_id, relations: List[Permission]):
        super().__init__(relations)
        self.user_id = group_id
        self.user_type = "group"

    async def add(self, mongo: DB):

        if not await id_exists(mongo.groups, self.user_id):
            raise ResourceNotFoundError

        update = {"permissions": {}}

        permission_dict = {permission.name: True for permission in self.relations}

        permissions = await get_one_field(
            mongo.groups, "permissions", {"_id": self.user_id}
        )

        update["permissions"] = {
            **permissions,
            **permission_dict,
        }

        async with mongo.create_session() as session:
            await mongo.groups.update_one(
                    {"_id": self.user_id},
                    {"$set": update},
                    session=session,
                )

            await update_member_users(mongo, self.user_id, session=session)

    async def remove(self, mongo: DB):

        if not await id_exists(mongo.groups, self.user_id):
            raise ResourceNotFoundError

        update = {"permissions": {}}

        permission_dict = {permission.name: False for permission in self.relations}

        permissions = await get_one_field(
            mongo.groups, "permissions", {"_id": self.user_id}
        )

        update["permissions"] = {
            **permissions,
            **permission_dict,
        }

        async with mongo.create_session() as session:
            await mongo.groups.update_one(
                    {"_id": self.user_id},
                    {"$set": update},
                    session=session,
                )

            await update_member_users(mongo, self.user_id, session=session)


class UserPermissions(BaseRelationship):
    def __init__(self, user_id, relations):
        super().__init__(relations)
        self.user_id = user_id
