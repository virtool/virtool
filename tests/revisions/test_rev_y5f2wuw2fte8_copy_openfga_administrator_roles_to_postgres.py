"""copy openfga administrator roles to postgres

Revision ID: y5f2wuw2fte8
Date: 2025-09-15 21:35:21.415101

"""

from collections.abc import Awaitable, Callable, Sequence

import pytest
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession
from syrupy import SnapshotAssertion

from assets.revisions.rev_y5f2wuw2fte8_copy_openfga_administrator_roles_to_postgres import (
    upgrade,
)
from virtool.authorization.relationships import AdministratorRoleAssignment
from virtool.migration import MigrationContext
from virtool.models.roles import AdministratorRole


class TestUpgrade:
    """Verify that administrator roles are correctly copied from OpenFGA to Postgres."""

    fetch_users: Callable[[], Awaitable[Sequence[dict]]]

    @pytest.fixture(autouse=True)
    async def setup(self, ctx: MigrationContext, static_time):
        """Ensure the environment is set up for each test."""
        # Create the enum type if it doesn't exist
        async with ctx.pg.begin() as conn:
            # Check if type exists
            result = await conn.execute(
                text("SELECT 1 FROM pg_type WHERE typname = 'administratorrole'")
            )
            if not result.fetchone():
                await conn.execute(
                    text(
                        "CREATE TYPE administratorrole AS ENUM ('full', 'settings', 'spaces', 'users', 'base')"
                    )
                )
            await conn.commit()

        # Create the user tables using raw SQL to avoid enum mapping issues
        async with ctx.pg.begin() as conn:
            await conn.execute(
                text("""
                CREATE TABLE IF NOT EXISTS users (
                    id SERIAL PRIMARY KEY,
                    handle VARCHAR NOT NULL UNIQUE,
                    legacy_id VARCHAR UNIQUE,
                    last_password_change TIMESTAMP NOT NULL,
                    password BYTEA,
                    settings JSONB NOT NULL,
                    administrator_role administratorrole
                )
            """)
            )
            await conn.commit()

        async def fetch_users() -> Sequence[dict]:
            async with AsyncSession(ctx.pg) as session:
                result = await session.execute(
                    text(
                        "SELECT id, handle, legacy_id, administrator_role FROM users ORDER BY id"
                    )
                )
                return [
                    {
                        "id": r.id,
                        "handle": r.handle,
                        "legacy_id": r.legacy_id,
                        "administrator_role": r.administrator_role,
                    }
                    for r in result.fetchall()
                ]

        self.fetch_users = fetch_users

    async def test_upgrade(
        self,
        ctx: MigrationContext,
        snapshot: SnapshotAssertion,
        static_time,
    ) -> None:
        """Verify administrator roles are copied from OpenFGA to Postgres."""
        # Create users in Postgres using raw SQL
        async with AsyncSession(ctx.pg) as session:
            await session.execute(
                text("""
                INSERT INTO users (handle, legacy_id, last_password_change, password, settings)
                VALUES
                    ('full_admin', 'full_admin_1', :time, :password, :settings),
                    ('spaces_admin', 'spaces_admin_1', :time, :password, :settings),
                    ('regular_user', 'user_1', :time, :password, :settings),
                    ('no_legacy_id', NULL, :time, :password, :settings)
            """),
                {
                    "time": static_time.datetime,
                    "password": b"password",
                    "settings": "{}",
                },
            )
            await session.commit()

        # Set up administrator roles in OpenFGA
        await ctx.authorization.add(
            AdministratorRoleAssignment("full_admin_1", AdministratorRole.FULL),
            AdministratorRoleAssignment("spaces_admin_1", AdministratorRole.SPACES),
        )

        await upgrade(ctx)

        assert await self.fetch_users() == snapshot

    async def test_upgrade_no_roles(
        self,
        ctx: MigrationContext,
        snapshot: SnapshotAssertion,
        static_time,
    ) -> None:
        """Verify users without administrator roles are left unchanged."""
        # Create users in Postgres without any administrator roles in OpenFGA
        async with AsyncSession(ctx.pg) as session:
            await session.execute(
                text("""
                INSERT INTO users (handle, legacy_id, last_password_change, password, settings)
                VALUES
                    ('user_1', 'user_1', :time, :password, :settings),
                    ('user_2', 'user_2', :time, :password, :settings)
            """),
                {
                    "time": static_time.datetime,
                    "password": b"password",
                    "settings": "{}",
                },
            )
            await session.commit()

        await upgrade(ctx)

        assert await self.fetch_users() == snapshot

    async def test_upgrade_existing_role(
        self,
        ctx: MigrationContext,
        snapshot: SnapshotAssertion,
        static_time,
    ) -> None:
        """Verify users with existing roles in Postgres are updated correctly."""
        # Create user in Postgres with existing administrator role
        async with AsyncSession(ctx.pg) as session:
            await session.execute(
                text("""
                INSERT INTO users (handle, legacy_id, last_password_change, password, settings, administrator_role)
                VALUES ('admin_user', 'admin_1', :time, :password, :settings, 'base')
            """),
                {
                    "time": static_time.datetime,
                    "password": b"password",
                    "settings": "{}",
                },
            )
            await session.commit()

        # Set up different administrator role in OpenFGA
        await ctx.authorization.add(
            AdministratorRoleAssignment("admin_1", AdministratorRole.FULL),
        )

        await upgrade(ctx)

        assert await self.fetch_users() == snapshot
