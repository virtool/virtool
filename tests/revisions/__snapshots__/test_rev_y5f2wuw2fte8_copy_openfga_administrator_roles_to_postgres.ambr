# serializer version: 1
# name: TestUpgrade.test_upgrade
  list([
    dict({
      'administrator_role': 'full',
      'handle': 'full_admin',
      'id': 1,
      'legacy_id': 'full_admin_1',
    }),
    dict({
      'administrator_role': 'spaces',
      'handle': 'spaces_admin',
      'id': 2,
      'legacy_id': 'spaces_admin_1',
    }),
    dict({
      'administrator_role': None,
      'handle': 'regular_user',
      'id': 3,
      'legacy_id': 'user_1',
    }),
    dict({
      'administrator_role': None,
      'handle': 'no_legacy_id',
      'id': 4,
      'legacy_id': None,
    }),
  ])
# ---
# name: TestUpgrade.test_upgrade_existing_role
  list([
    dict({
      'administrator_role': 'full',
      'handle': 'admin_user',
      'id': 1,
      'legacy_id': 'admin_1',
    }),
  ])
# ---
# name: TestUpgrade.test_upgrade_no_roles
  list([
    dict({
      'administrator_role': None,
      'handle': 'user_1',
      'id': 1,
      'legacy_id': 'user_1',
    }),
    dict({
      'administrator_role': None,
      'handle': 'user_2',
      'id': 2,
      'legacy_id': 'user_2',
    }),
  ])
# ---
